browser/onboarding/onboarding_tab_helper.h

```diff
@@ -0,0 +1,39 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_ONBOARDING_ONBOARDING_TAB_HELPER_H_
+#define BRAVE_BROWSER_ONBOARDING_ONBOARDING_TAB_HELPER_H_
+
+#include <string>
+
+#include "brave/browser/ui/brave_shields_data_controller.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+class OnboardingTabHelper
+    : public content::WebContentsObserver,
+      public content::WebContentsUserData<OnboardingTabHelper> {
+ public:
+  static void MaybeCreateForWebContents(content::WebContents* contents);
+  OnboardingTabHelper(const OnboardingTabHelper&) = delete;
+  OnboardingTabHelper& operator=(const OnboardingTabHelper&) = delete;
+  ~OnboardingTabHelper() override;
+
+ private:
+  friend class content::WebContentsUserData<OnboardingTabHelper>;
+
+  explicit OnboardingTabHelper(content::WebContents* web_contents);
+
+  // content::WebContentsObserver
+  void DidStopLoading() override;
```

Comment:
Do we need to be paranoid and override `WebContentsDestroyed` and remove the OnboardingTabHelper itself?

```c++
if (g_browser_process->local_state()->GetBoolean(onboarding::prefs::kOnboardingIsShieldsHighlighted))
   web_contents()->RemoveUserData(UserDataKey());
```

cc @petemill @sangwoo108

Reply:
> Do we need to be paranoid and override WebContentsDestroyed

Yeah, I think override WebContentsDestroyed and unobserve the contents would be better as we usaually want the source object doesn't have any observers on it's destruction. But deleting itself itself on that callback might not be necessary.

As for the first run case, I like your idea. We could the attached tab helpers on the preference change event.

Reply:
> unobserve the contents would be better

Do we need to keep a cache of active web contents here?

Reply:
I don't think so. You should be able to gett valid web contents with `web_contents()` at that point.

Reply:
I guess there isn't way to "unobserve" but I will have to call `Observe(nullptr);` ?

Reply:
Yes, that's correct!